File: trifecta-ios/Trifecta/Core/Networking/SshClient.swift
import Foundation

extension T3Client {
    func sshListHosts() async throws -> [SshHostProfile] {
        guard let value = try await request(method: "ssh.listHosts", payload: [String: Any]()) else {
            throw T3Error.decodingFailed("Empty ssh.listHosts response")
        }
        return try SshHostProfile.decodeList(from: value)
    }

    func sshAddHost(label: String,
                    hostname: String,
                    port: Int,
                    username: String,
                    authMethod: SshAuthMethod) async throws -> SshHostProfile {
        let payload: [String: Any] = [
            "label": label,
            "hostname": hostname,
            "port": port,
            "username": username,
            "authMethod": authMethod.rawValue
        ]
        guard let value = try await request(method: "ssh.addHost", payload: payload) else {
            throw T3Error.decodingFailed("Empty ssh.addHost response")
        }
        let data = try JSONSerialization.data(withJSONObject: value)
        return try JSONDecoder().decode(SshHostProfile.self, from: data)
    }

    func sshRemoveHost(hostId: String) async throws {
        _ = try await request(method: "ssh.removeHost", payload: ["hostId": hostId])
    }

    func sshOpenSession(hostId: String, cols: Int = 80, rows: Int = 24) async throws -> SshOpenSessionResult {
        let payload: [String: Any] = ["hostId": hostId, "cols": cols, "rows": rows]
        guard let value = try await request(method: "ssh.openSession", payload: payload) else {
            throw T3Error.decodingFailed("Empty ssh.openSession response")
        }
        return try SshOpenSessionResult.decode(from: value)
    }

    func sshSendInput(sessionId: String, data: String) async throws {
        _ = try await request(method: "ssh.sendInput", payload: ["sessionId": sessionId, "data": data])
    }

    func sshResize(sessionId: String, cols: Int, rows: Int) async throws {
        _ = try await request(method: "ssh.resize", payload: [
            "sessionId": sessionId,
            "cols": cols,
            "rows": rows
        ])
    }

    func sshConfirmHostKey(sessionId: String,
                           fingerprintSha256: String,
                           approve: Bool,
                           remember: Bool) async throws -> SshSessionSnapshot {
        let payload: [String: Any] = [
            "sessionId": sessionId,
            "fingerprintSha256": fingerprintSha256,
            "decision": approve ? "approve" : "reject",
            "remember": remember
        ]
        guard let value = try await request(method: "ssh.confirmHostKey", payload: payload) else {
            throw T3Error.decodingFailed("Empty ssh.confirmHostKey response")
        }
        return try SshSessionSnapshot.decode(from: value)
    }

    func sshCloseSession(sessionId: String) async throws {
        _ = try await request(method: "ssh.closeSession", payload: ["sessionId": sessionId])
    }

    struct ShellProfileSetupResult {
        let shellProfile: String
        let alreadyPresent: Bool
    }

    func sshSetupShellProfile() async throws -> ShellProfileSetupResult {
        guard let value = try await request(method: "ssh.setupShellProfile", payload: [String: Any]()) as? [String: Any] else {
            throw T3Error.decodingFailed("Empty ssh.setupShellProfile response")
        }
        guard let shellProfile = value["shellProfile"] as? String,
              let alreadyPresent = value["alreadyPresent"] as? Bool else {
            throw T3Error.decodingFailed("Invalid ssh.setupShellProfile response")
        }
        return ShellProfileSetupResult(shellProfile: shellProfile, alreadyPresent: alreadyPresent)
    }

    func subscribeSshTerminal(sessionId: String,
                              onEvent: @escaping (SshTerminalEvent) -> Void) async throws -> StreamSubscription {
        try await subscribe(method: "subscribeSshTerminal", payload: ["sessionId": sessionId]) { value in
            do {
                let event = try SshTerminalEvent.decode(from: value)
                onEvent(event)
            } catch {
                NSLog("Failed to decode SSH terminal event: \(error)")
            }
        }
    }
}

File: trifecta-android/app/src/main/java/com/belweave/trifecta/core/networking/SshClient.kt
package com.belweave.trifecta.core.networking

import android.util.Log
import com.belweave.trifecta.core.models.SshAuthMethod
import com.belweave.trifecta.core.models.SshHostProfile
import com.belweave.trifecta.core.models.SshOpenSessionResult
import com.belweave.trifecta.core.models.SshSessionSnapshot
import com.belweave.trifecta.core.models.SshTerminalEvent
import com.belweave.trifecta.core.models.asArrayOrNull
import com.belweave.trifecta.core.models.asObjectOrNull
import com.belweave.trifecta.core.models.booleanOrNull
import com.belweave.trifecta.core.models.stringOrNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

data class ShellProfileSetupResult(
    val shellProfile: String,
    val alreadyPresent: Boolean
)

fun interface SshTerminalEventListener {
    fun onEvent(event: SshTerminalEvent)
}

suspend fun T3Client.sshListHosts(): List<SshHostProfile> {
    val value = request("ssh.listHosts", JsonObject(emptyMap()))
        ?: throw T3Error.DecodingFailed("Empty ssh.listHosts response")
    val obj = value.asObjectOrNull() ?: throw T3Error.DecodingFailed("ssh.listHosts response was not an object")
    return obj["hosts"]
        ?.asArrayOrNull()
        ?.mapNotNull { it.asObjectOrNull()?.let(SshHostProfile.Companion::fromJson) }
        ?: obj["hosts"]
            ?.asObjectOrNull()
            ?.values
        ?.mapNotNull { it.asObjectOrNull()?.let(SshHostProfile.Companion::fromJson) }
        ?: emptyList()
}

suspend fun T3Client.sshAddHost(
    label: String,
    hostname: String,
    port: Int,
    username: String,
    authMethod: SshAuthMethod
): SshHostProfile {
    val payload = buildJsonObject {
        put("label", label)
        put("hostname", hostname)
        put("port", port)
        put("username", username)
        put("authMethod", authMethod.raw)
    }
    val value = request("ssh.addHost", payload)
        ?: throw T3Error.DecodingFailed("Empty ssh.addHost response")
    val obj = value.asObjectOrNull() ?: throw T3Error.DecodingFailed("ssh.addHost response was not an object")
    return SshHostProfile.fromJson(obj) ?: throw T3Error.DecodingFailed("Could not decode ssh host profile")
}

suspend fun T3Client.sshRemoveHost(hostId: String) {
    request("ssh.removeHost", buildJsonObject { put("hostId", hostId) })
}

suspend fun T3Client.sshOpenSession(hostId: String, cols: Int = 80, rows: Int = 24): SshOpenSessionResult {
    val payload = buildJsonObject {
        put("hostId", hostId)
        put("cols", cols)
        put("rows", rows)
    }
    val value = request("ssh.openSession", payload)
        ?: throw T3Error.DecodingFailed("Empty ssh.openSession response")
    val obj = value.asObjectOrNull() ?: throw T3Error.DecodingFailed("ssh.openSession response was not an object")
    return SshOpenSessionResult.fromJson(obj) ?: throw T3Error.DecodingFailed("Could not decode ssh session")
}

suspend fun T3Client.sshSendInput(sessionId: String, data: String) {
    request("ssh.sendInput", buildJsonObject {
        put("sessionId", sessionId)
        put("data", data)
    })
}

suspend fun T3Client.sshResize(sessionId: String, cols: Int, rows: Int) {
    request("ssh.resize", buildJsonObject {
        put("sessionId", sessionId)
        put("cols", cols)
        put("rows", rows)
    })
}

suspend fun T3Client.sshConfirmHostKey(
    sessionId: String,
    fingerprintSha256: String,
    approve: Boolean,
    remember: Boolean
): SshSessionSnapshot {
    val payload = buildJsonObject {
        put("sessionId", sessionId)
        put("fingerprintSha256", fingerprintSha256)
        put("decision", if (approve) "approve" else "reject")
        put("remember", remember)
    }
    val value = request("ssh.confirmHostKey", payload)
        ?: throw T3Error.DecodingFailed("Empty ssh.confirmHostKey response")
    val obj = value.asObjectOrNull() ?: throw T3Error.DecodingFailed("ssh.confirmHostKey response was not an object")
    return SshSessionSnapshot.fromJson(obj) ?: throw T3Error.DecodingFailed("Could not decode ssh session snapshot")
}

suspend fun T3Client.sshCloseSession(sessionId: String) {
    request("ssh.closeSession", buildJsonObject { put("sessionId", sessionId) })
}

suspend fun T3Client.sshSetupShellProfile(): ShellProfileSetupResult {
    val value = request("ssh.setupShellProfile", JsonObject(emptyMap()))
        ?: throw T3Error.DecodingFailed("Empty ssh.setupShellProfile response")
    val obj = value.asObjectOrNull() ?: throw T3Error.DecodingFailed("ssh.setupShellProfile response was not an object")
    val shellProfile = obj["shellProfile"].stringOrNull()
        ?: throw T3Error.DecodingFailed("Missing shellProfile in ssh.setupShellProfile response")
    val alreadyPresent = obj["alreadyPresent"].booleanOrNull()
        ?: throw T3Error.DecodingFailed("Missing alreadyPresent in ssh.setupShellProfile response")
    return ShellProfileSetupResult(shellProfile = shellProfile, alreadyPresent = alreadyPresent)
}

suspend fun T3Client.subscribeSshTerminal(
    sessionId: String,
    listener: SshTerminalEventListener
): StreamSubscription {
    return subscribe("subscribeSshTerminal", buildJsonObject { put("sessionId", sessionId) }) { value ->
        val event = SshTerminalEvent.fromJson(value)
        if (event != null) {
            listener.onEvent(event)
        } else {
            Log.w("T3Client", "Discarded SSH terminal event: $value")
        }
    }
}
